Яценко Ирина

lholypeachy opened the pull request:
@LevShisterov

lholypeachy committed:
Начальное проектирование

lholypeachy committed:
Реализация сохранения облака тегов в png файл

lholypeachy committed:
Добавлен декоратор для TextProcessor

LevShisterov commented on TagCloudGenerator/Properties/launchSettings.json:
```diff
@@ -0,0 +1,8 @@
+{
```
launchSettings.json обычно не коммитят, так как он зависит от окружения на машине разработчика.

LevShisterov commented on TagCloudGenerator/ITextProcessor.cs:
```diff
@@ -0,0 +1,7 @@
+{
+    public interface ITextProcessor
+    {
+        string[] ProcessText(string[] file);
```
`string[] file` - "массив текстовых файлов"? Сложно понять что ожидается на входе.

LevShisterov commented on TagCloudGenerator/TagCloudDrawer.cs:
```diff
@@ -0,0 +1,54 @@
+    public class TagCloudDrawer
+    {
+        public VisualizingSettings settings;
+        public void DrawWordsCloud(string filePath)
```
А если передать null?

LevShisterov commented on TagCloudGenerator/TagCloudDrawer.cs:
```diff
@@ -0,0 +1,54 @@
+
+        private void Draw(string[] text)
+        {
+            var center = new Point(500, 500);
+            var distributor = new Spiral(1,center, 0.1);
+            var layouter = new CircularCloudLayouter(center, distributor);
+            var brush = new SolidBrush(Color.Aqua);
+            var font = new Font("Arial", 24);
```
Вот это всё должно быт или в объекте с настройками или зависимостями и собираться с помощью контейнера.

LevShisterov commented on TagCloudGenerator/TagCloudDrawer.cs:
```diff
@@ -0,0 +1,54 @@
+        public VisualizingSettings settings;
+        public void DrawWordsCloud(string filePath)
+        {
+            var tagCloudDrawer = new TagCloudDrawer();
```
Тут я не очень понял, TagCloudDrawer создает ещё одну версию себя, зачем?

LevShisterov commented on TagCloudGenerator/TagCloudDrawer.cs:
```diff
@@ -0,0 +1,54 @@
+            var brush = new SolidBrush(Color.Aqua);
+            var font = new Font("Arial", 24);
+
+            var bitmap = new Bitmap(1000, 1000);
+            var graphics = Graphics.FromImage(bitmap);
```
Это IDispasable типы - надо диспозить, после использования.

LevShisterov commented on TagCloudGenerator/TagCloudDrawer.cs:
```diff
@@ -0,0 +1,54 @@
+          
+            for (var i = 0; i < text.Length; i++)
+            {
+                SizeF size = graphics.MeasureString(text[i], font);
+                var rect = layouter.PutNextRectangle(size.ToSize());
+
+                if (i == 0 )
+                    distributor.centerOnPoint = true;
+
+                graphics.DrawString(text[i], font, brush, rect.X, rect.Y);
```
Если смотреть на это и на `Console.WriteLine()` чуть выше, то кажется что класс делает два дела. Может его надо разделить?
И, `bitmap.Save` можно считать третьим делом. Например, если будет GUI клиент и надо не сохранять в файл, а рисовать на экран?

LevShisterov commented on TagCloudGenerator/TextProcessorWrapper.cs:
```diff
@@ -0,0 +1,16 @@
+﻿namespace TagCloudGenerator
+{
+    public abstract class TextProcessorWrapper : ITextProcessor
```
Зачем этот класс?

LevShisterov replied:
Базовая реализация метода ProcessText этого класса нигде не вызывается, ссылок на этот тип я не вижу. Больше класс ничего не делает, может он не нужен?

LevShisterov commented on TagCloudGenerator/TextProcessor.cs:
```diff
@@ -0,0 +1,15 @@
+    {
+        public string[] ProcessText(string[] file)
+        {
+            for (var i = 0; i < file.Length; i++)
```
Может `foreach`?
И вообще, можно было бы принимать `IEnumerable` и его же возвращать. Зачем ограничивать себя массивами?

LevShisterov commented on TagCloudGenerator/WordCounter.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public class WordCounter
+    {
+        public Dictionary<string, int> words;
```
Не видно необходимости в этом поле.

LevShisterov commented on TagCloudGeneratorTest/TagCloudGeneratorTests.cs:
```diff
@@ -0,0 +1,42 @@
+                result += (fileText[i] + Environment.NewLine);
+            }
+
+            result.Should().Be("��������\r\n������\r\n�����\r\n��\r\n�����");
```
Кодировка в файле не utf8 (посмотри в ПР на гитбабе).

LevShisterov commented on TagCloudGeneratorTest/TagCloudGeneratorTests.cs:
```diff
@@ -0,0 +1,42 @@
+        [Test]
+        public void WhenPassWordsInUppercase_ShouldReturnWordsInLowerCase()
+        {
+            var filePath = @"C:\Users\lholy\Documents\GitHub\di\TagCloudGeneratorTest\TestsData\test1.txt";
+            var file = File.ReadAllLines(filePath);
```
Нет смысла читать из файла, если TextProcessor.ProcessText всё равно принимает на вход строки.

LevShisterov commented on TagsCloudVisualization/Program.cs:
```diff
@@ -0,0 +1,16 @@
+        public const int ImageWidth = 1000;
+        public const int ImageHeight = 1000;
+
+        public static void Main(string[] args)
```
Кажется тип проекта TagsCloudVisualization  должен быть Library, т.е. не предполагать прямого запуска и не иметь Main метода.

LevShisterov commented on TagsCloudVisualization/VisualizingSettings.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+        public VisualizingSettings(string imageName, Size imageSize)
+        {
+            ImageName = imageName;
```
Кажется, отрисовка облака в bitmap не предполагает что для этого нужно название.

LevShisterov commented on TagsCloudVisualization/TagCloudRenderer.cs:
```diff
@@ -0,0 +1,106 @@
+        private void CustomizeGraphics(VisualizingSettings settings)
+        {
+            pen = new Pen(settings.PenColor);
+            graphics = Graphics.FromImage(bitmap);
```
graphics - IDisposable объект.

LevShisterov commented on TagsCloudVisualization/TagCloudRenderer.cs:
```diff
@@ -0,0 +1,106 @@
+                return;
+            }
+
+            bitmap = new Bitmap(unscaledImageSize.Width, unscaledImageSize.Height);
```
bitmap - IDisposable объект.

lholypeachy committed:
Дополнен класс Options

Добавлена возможность задать цвета, шрифт, размер изображения и возможность использования одного алгоритма с разными настройками.
Изменена кодировка файла.

lholypeachy committed:
Реализация пунктов на перспективу

1. Поддерживать разные форматы файлов
2. поддерживать разные форматы изображений.

lholypeachy committed:
Реализована сборка зависимостей

lholypeachy committed:
Реализовано исключение "скучных" слов

lholypeachy committed:
Исправление ошибок

lholypeachy committed:
Добавлена возможность читать pdf файлы и новые тестовые случаи

LevShisterov commented on TagCloudGenerator/TagCloudDrawer.cs:
```diff
@@ -0,0 +1,109 @@
+            }
+
+           return bitmap;
+        }  
```
В этом методе много IDispasable объектов. Graphics, Font, Brush, Bitmap - внутри используют Handle из системы, если их быстро, много насоздавать, могут кончиться. Надо диспозить, после использования, чтобы освободить поскорее.

LevShisterov commented on TagCloudGeneratorTest/TagCloudGeneratorTests.cs:
```diff
@@ -0,0 +1,144 @@
+            foreach (var word in text)
+                result += (word + Environment.NewLine);
+
+            result.Should().Be("cloud\r\ntags\r\n");
```
Кажется что есть более простые способы проверить содержимое массива, чем склеивать его в строку.

LevShisterov commented on TagCloudGeneratorTest/TagCloudGeneratorTests.cs:
```diff
@@ -0,0 +1,144 @@
+        {
+            var currentBitmap = GetCurrentImage();
+
+            var pathToImage = "C:\\Users\\lholy\\Documents\\GitHub\\di\\TagCloudGeneratorTest\\TestsData\\ForTests.png";
```
Тут уж как-то совсем путь захардкожен. другому разработчику придется править тест чтобы он заработал у него.

LevShisterov commented on TagsCloudVisualization/TagCloudRenderer.cs:
```diff
@@ -0,0 +1,112 @@
+            graphics.DrawRectangles(pen, rectangles);
+
+            var coefficient = GetScaleCoefficients(unscaledImageSize, settings.ImageSize);
+            graphics.ScaleTransform(coefficient.sx, coefficient.sy);
```
Так не получится разное масштабирование по x и y?

LevShisterov commented on TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,52 @@
+        private readonly Cloud cloud;
+        private readonly IPointDistributor distributor;
+
+        public CircularCloudLayouter(Point center, IPointDistributor type)
```
Тут, есть два center значения. Одно передается непосредственно, и одно есть в Spiral.
Может можно не передавать сюда center? Просто если прямоугольников еще нет, можно сместить первый на половинку высоты и ширины.

LevShisterov commented on TagsCloudVisualization/TagCloudRenderer.cs:
```diff
@@ -0,0 +1,112 @@
+            var unscaledImageSize = GetImageSizeWithRealSizeRectangles(rectangles, smallestSizeOfRectangles);
+
+            if (unscaledImageSize.Height <= settings.ImageSize.Height &&
+                unscaledImageSize.Width <= settings.ImageSize.Width)
```
После этого if идет довольно много одинакового кода, можно проверив это условие сделать какой-нибудь признак что надо отмаштабировать, рзамеры взять максимальные, код может стать проще.

LevShisterov commented on TagsCloudVisualization/TagCloudRenderer.cs:
```diff
@@ -0,0 +1,112 @@
+
+            using (bitmap = new Bitmap(unscaledImageSize.Width, unscaledImageSize.Height))
+            {
+                CustomizeGraphics(settings);
```
В CustomizeGraphics объект в поле graphics диспозится, далее доступ к нему в DrawScaleCloud должен падать.

LevShisterov commented on TagsCloudVisualization/TagCloudRenderer.cs:
```diff
@@ -0,0 +1,112 @@
+{
+    public class TagCloudRenderer
+    {
+        private Bitmap bitmap;
+        private Graphics graphics;
+        private Pen pen;
```
Искуственные поля, вызыванные разделением методов. Можно параметрами передать.
И ещё, видя такие поля (типа IDisposable) возникает вопрос: а где Dispose метод для освобождения полей?
Так что, проще без них.

LevShisterov commented on TagCloudGenerator/TagCloudDrawer.cs:
```diff
@@ -0,0 +1,109 @@
+                throw new ArgumentNullException(nameof(filePath));
+           
+            var words = textReader.ReadTextFromFile(filePath);        
+            var deleteBoringWords = new BoringWordsTextProcessor(textProcessor);
```
В классе TagCloudDrawer, в методе DrawWordsCloud имеется `new BoringWordsTextProcessor`.
Что это означает? Что класс TagCloudDrawer жестко зависит от BoringWordsTextProcessor, его не получится использовать без BoringWordsTextProcessor. Пользователь класса TagCloudDrawer не может никак заменить BoringWordsTextProcessor.
Как можно решить эту проблему?

lholypeachy committed:
Исправление ошибок

lholypeachy committed:
Корректировка ITextProcessor и изменение расчета коэффициента масштабирования

lholypeachy committed:
Преобразование TextProcessors в коллекцию

lholypeachy committed:
Разбиение процессов обработки текста на соответствующие классы

lholypeachy committed:
Рефакторинг кода
